annotator-core/src/main/java/edu/ucr/cs/riple/core/Config.java

```diff
@@ -701,4 +719,65 @@ public void write(Path path) {
       }
     }
   }
+
+  /** Responsible for handling offset changes in source file. */
+  public static class OffsetHandler {
+    /** Map of file paths to list offset changes. */
+    private final Map<Path, List<OffsetChange>> contents;
+
+    public OffsetHandler() {
+      contents = new HashMap<>();
+    }
+
+    /**
+     * Gets the original offset according to existing offset changes.
+     *
+     * @param path Path to source file.
+     * @param offset Given offset.
+     * @return Original offset.
+     */
+    public int getOriginalOffset(Path path, int offset) {
+      return OffsetChange.getOriginalOffset(offset, contents.getOrDefault(path, List.of()));
+    }
+
+    /**
+     * Updates given offsets with given new offset changes.
+     *
+     * @param newOffsets Given new offset changes.
+     */
+    public void updateOffset(Set<FileOffsetStore> newOffsets) {
+      newOffsets.forEach(
+          store -> {
```

Comment:
I didn't fully understand this code, but just to answer to your question, it is possible that both existing collection and the new incoming collection be sorted and summarized but the union of them require summarization. Also I am not sure what the first line means (`store` is not defined). Please let me know if I missed anything thank you.

Reply:
Ah, I think I missed something here, in that I thought `newOffsets` was a set of cumulative updates applying sequentially on top of the same `existingOffsetChanges`, but they are actually partitioned by `.getPath()`, correct? As in, there shouldn't be two stores in `Set<FileOffsetStore> newOffsets` with the same path? If so, then partitioning the iteration like you are already doing makes sense, it's just updating the corresponding path in `contents` for each distinct path in `newOffsets`.

I think if that's the case, then it would be a lot clearer for the argument here to be `Map<Path, FileOffsetStore>`.

Reply:
Sure, sounds good. Thank you for the suggestion. I also moved the logic to `FileOffsetStore` since we keep the map to them. Looks cleaner now. a11ed6f19d319dd07fa2e57160ed0f5576234fc2